// containers/phone/phone.js
import React, {Component} from 'react'
import {connect} from 'react-redux'
import {fetchPhoneById} from "actions";
import {getPhoneById} from "../../selectors";



class Phone extends Component{
    componentDidMount() {
        console.log('phone', this.props);
        this.props.fetchPhoneById(this.props.match.params.id)
    }
    renderFields () {
        const {phone} = this.props;
        const list = {
            'cpu': phone.cpu,
            'camera': phone.camera,
            'size': phone.size,
            'weight': phone.weight,
            'display': phone.display,
            'battery': phone.battery,
            'memory': phone.memory
        }; //создание обьекта из нужных свойств объекта phone
        const arrColumnFields = Object.entries(list);//создание массива из обьекта list

        return arrColumnFields.map(([key, value]) => {
            return (
                <div className='column' key={key}>
                    <div className='ab-details-title'>
                        <p>{key}</p>
                    </div>
                    <div className='ab-details-info'>
                        {value}
                    </div>
                </div>
            )
        })
    }


    renderContent () {
        const {phone} = this.props;
        return(
            <div className='thumbnail'>
                <div className='row'>
                    <div className='col-6'>
                        <img className="img-thumbnail" src={phone.image} alt={phone.name}/>
                    </div>
                    <div className='col-6'>
                        {this.renderFields()}
                    </div>
                </div>
                <div className="caption-full">
                    <h4 className='pull-right'>$ {phone.price}</h4>
                    <h4>{phone.name}</h4>
                    <p>{phone.description}</p>
                </div>
            </div>
        )
    }
    renderSideBar() {
        return(
            <div>sidebar</div>
        )
    }
    render(){
        console.log('phone props:', this.props.phone);
        const {phone} = this.props;
        return (
            <div className='view-container'>
                <div className='container'>
                    <div className='row'>
                        <div className='col-9'>
                            {phone && this.renderContent()}
                        </div>
                        <div className='col-3'>
                            {phone && this.renderSideBar()}
                        </div>
                    </div>
                </div>
            </div>
        )
    }
}
const mapStateToProps = state => {
    return{
        phone: getPhoneById(state, state.phonePage.id)
    }
}
const mapDispatchToProps  = {
    fetchPhoneById
}

export default connect(mapStateToProps, mapDispatchToProps)(Phone);
